lgc/patch/PatchWorkarounds.cpp

```diff
@@ -0,0 +1,141 @@
+ *
+ *  All rights reserved. This notice is intended as a precaution against inadvertent publication and does not imply
+ *  publication or any waiver of confidentiality. The year included in the foregoing notice is the year of creation of
+ *  the work.
+ *
+ **********************************************************************************************************************/
+/**
+ ***********************************************************************************************************************
+ * @file  PatchWorkarounds.cpp
+ * @brief LLPC source file: contains implementation of class lgc::PatchWorkarounds.
+ ***********************************************************************************************************************
+ */
+
+#include "PatchWorkarounds.h"
+#include "lgc/BuilderBase.h"
+#include "lgc/state/PipelineShaders.h"
+#include "lgc/state/PipelineState.h"
+#include "lgc/state/TargetInfo.h"
+#include "llvm/IR/Constants.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/IntrinsicsAMDGPU.h"
+#include "llvm/Support/Debug.h"
+
+#define DEBUG_TYPE "lgc-patch-workarounds"
+
+using namespace lgc;
+using namespace llvm;
+
+namespace lgc {
+
+// =====================================================================================================================
+// Define static members (no initializer needed as LLVM only cares about the address of ID, never its value).
+char PatchWorkarounds::ID;
+
+// =====================================================================================================================
+// Pass creator, creates the pass of LLVM patching operations for peephole optimizations.
+FunctionPass *createPatchWorkarounds() {
+  return new PatchWorkarounds();
+}
+
+PatchWorkarounds::PatchWorkarounds() : FunctionPass(ID) {
+}
+
+// =====================================================================================================================
+// Executes this LLVM pass on the specified LLVM function.
+//
+// @param [in/out] function : Function that we will add workarounds.
+bool PatchWorkarounds::runOnFunction(Function &function) {
+  LLVM_DEBUG(dbgs() << "Run the pass Patch-Workarounds\n");
+
+  m_pipelineState = getAnalysis<PipelineStateWrapper>().getPipelineState(function.getParent());
+  m_context = &function.getContext();
+  m_builder = std::make_unique<IRBuilder<>>(*m_context);
+  m_processed.clear();
+
+  m_changed = false;
+
+  // At present we only apply a single workaround in this pass. Makes sense to check here an early exit the pass if not
+  // required
+  if (!m_pipelineState->getOptions().disableImageResourceCheck &&
+      m_pipelineState->getTargetInfo().getGpuWorkarounds().gfx10.waFixBadImageDescriptor)
+    visit(function);
+
+  return m_changed;
+}
+
+// =====================================================================================================================
+// Visits "call" instruction.
+//
+// @param callInst : The instruction
```

Comment:
Could you add comments that describe what this pass does and why? The existing comments only repeat what the source already says.

Reply:
Yes, although interestingly this was a cut-and-paste from elsewhere (but I agree with your comment).

Reply:
Done